src/moves.rs

```diff
@@ -79,23 +110,29 @@ impl Board {
         match judge.battle(attacking_words, defending_words) {
             Outcome::NoBattle => {}
             Outcome::DefenderWins => {
-                for word in attackers {
-                    for square in word {
-                        self.clear(square);
-                    }
-                }
+                let squares = attackers.into_iter().flat_map(|word| word.into_iter());
+                changes.extend(squares.flat_map(|square| {
+                    self.clear(square).map(|detail| Change {
+                        detail,
+                        change: ChangeAction::Defeated,
+                    })
+                }));
             }
             Outcome::AttackerWins(losers) => {
-                for defender_index in losers {
+                let squares = losers.into_iter().flat_map(|defender_index| {
                     let defender = defenders.get(defender_index).unwrap();
-                    for square in defender {
-                        self.clear(*square);
-                    }
-                }
+                    defender.into_iter()
+                });
+                changes.extend(squares.flat_map(|square| {
+                    self.clear(*square).map(|detail| Change {
+                        detail,
+                        change: ChangeAction::Defeated,
+                    })
+                }));
             }
         }
 
-        self.truncate();
+        changes.extend(self.truncate().into_iter());
```

Comment:
This `resolve_attack` takes a `&mut Vec<Change>` to avoid allocating a new Vec here.

Theoretically we could/should make the same change for the `truncate()` fn, but I haven't since it's in another file and that feels less than ideal.

Reply:
That's fair enough, but the distinction between `move.rs` and `board.rs` is pretty badly defined. `move.rs` is supposed to be game logic, whereas `board.rs` is supposed to be lower level board navigation logic (like finding adjacent squares etc).

But really, `truncate` should probably be in `move.rs` by that definition, I just kept in it `board.rs` to use the private `roots` property, which should arguably be visible.